fpc demangle assembly labels

partouf opened the pull request:
I couldn't find demangle functionality in the FPC sourcecode, so I decided to write it myself. I wrote a version in Pascal first, but I think that would be more trouble than worth to setup and use, so I rewrote it in javascript instead.

This supports both FPC 2.x and 3.x types of mangling.

RabsRincon commented:
Thanks for this! :)
There is one minor issue that would be great to resolve before merging. I don't know if you noticed, but there's a button in the compiler to enable/disable demangling, which I think it would be great to get working for Pascall as well. This button is currently disabled if the `demangler` compiler property is not set, so either
1) You compile a Pascal one and set it like you first tried
or
We leave it like this and just disallow disabling demangling for Pascal

What do you think could be done?

partouf commented:
@RabsRincon I noticed in live but hadn't done an update to my dev environment. I can setting demangler to a value is ok, because I override the postProcessAsm function and theres no "optOutput", so it doesn't actually call it from anywhere.

Added that to the commits.

RabsRincon commented:
Oh that is great, but is there now a way to not demangle the output? (Hint: the filters variable has a demangle member which tells you if the user wants to demangle the output)

partouf commented:
No you can now click the button and it leaves the assembly as is, because the postProcessAsm function isn't called in base-compiler.js based on
                return filters.demangle ? _.bind(self.postProcessAsm, self, result)() : result;

RabsRincon commented:
Oh right, I forgot about that, sorry!

So everything looks great, I'll give it a last look before merging.

Thanks for all you've done! :)

RabsRincon commented:
So, I've had this idea that would be great to have: We should refer to Pascal as Pascal internally, and not as fpc. That is, launching it as `make EXTRA_ARGS="--language Pascal"` and not `--language fpc`, having the properties be called `pascal.xxxx.properties`, having the subdomain be named `pascal` and moslty everywhere else.

Do you think it would be a good idea? You don't need to worry about the subdomain naming, of course, but would be great if once this PR is merged, you could do the renaming (That is, if you want to! Of course!)

RabsRincon commented on lib/compilers/fpc.js:
```diff
@@ -30,6 +30,7 @@ function compileFPC(info, env) {
     var compiler = new Compile(info, env);
     compiler.supportsOptOutput = false;
     compiler.originalPostProcess = compiler.postProcess;
+    compiler.symbolcache = [];
```
Is this used anywhere else? If it's not, I think it migh be better as a local variable to the compileFPC (IF! It's not reinitialised every time it's called, in which case I guess the only solution is to store it on the compiler..)

partouf replied:
Will see what I can do, wasn't too sure about the JS scopes in this case, but will experiment

RabsRincon commented:
I have noticed that I get a  `file.o` created on the repo root now. Can you please try to reproduce locally? It might be an unreleated change, but I suspect it's coming from the fpc backend

partouf commented:
@RabsRincon rebased and refactored and added line coloring and binary execution, so ... I'm sorry... and I'm still not super happy about the output, but it's better than how it was.

partouf commented:
includes #637

partouf commented:
Also concerning "file.o", I don't seem to get that file anywhere, so I'm not sure where that's coming from.

I am getting a ppas.sh file, I suspect FPC is doing that? But not sure what to do about. I could of course delete it after a build..

RabsRincon commented:
So the code now looks almost ready now. Last few issues are:
1) Have you seen https://github.com/mattgodbolt/compiler-explorer/pull/632#issuecomment-346224504? I think we could get that change in this PR if you want
2) Could you disable the execution in the amazon enviroment? It should be a simple. I think you just have to add `supportsExecution=false` in its properties file (Test that it indeed disallows execution by testing in local)
3) As @jaredwy mentioned, I'm not too sure about the global variable for the `usedStg` being there (The cache ones are fine tho, but you might want to use the LRUCache module (But that's optional, I'll approve even if you don't do it :) )) What happens when: 
    1) One person A compiles such that the value is true
    2) Another person B compiles. Value is set to false
    3) A's compilation result arrives
Is A's value true or false? (My guess is that it is indeed false) If you can confirm that it's true that would be great, but if not, try not to use a global variable. Maybe pass the selected option from the filters object/user options/another relevant variable?

Sorry to be such a pain in the neck! But you're doing great, thanks for your effort with getting this into CE! 

:)

partouf commented:
@RabsRincon I think for #2 this is done through compiler-explorer.amazon.properties? I don't see it on any other language properties file, but it's still disabled for the live website

RabsRincon commented:
You're right... I'm not the sharpest one today, sorry!

partouf commented:
While I'm renaming, I'm also gonna see if I can put more into the properties file instead of having it in the code, might be a while

mattgodbolt commented:
Thanks both for taking the time to improve this!

RabsRincon commented:
@partouf Good to hear! Let us know if you need any help in the processs

mattgodbolt commented:
Lots of changes there: cool! Let us know when it's ready for another look :)

RabsRincon commented:
I've reviewed the changes, and besides the addition to the asm.js file (Which is not wrong! I just wish there was a more self contained solution) everything looks great! So do let us know when this is ready to get merged
:)

partouf commented:
The change to asm.js was something I was mulling over these last few days, but it seems at the moment this is the least intrusive way of doing it. The way asm.js is currently set up is not very friendly to passing in new things.

So for now, I'd say this is it.

I think compiler-explorer-image needs changes to reflect these, right?

RabsRincon commented:
In that case, if you're happy with the state of the PR, we can merge!

It did need some changes, but have been done in advance

partouf commented:
ok cool

mattgodbolt commented:
There's still one `compiler-explorer-image` change that is needed: the `--language Pascal` part. Once we're ready to merge I'll orchestrate the changes together! Thanks :)

mattgodbolt commented on lib/compilers/pascal.js:
```diff
@@ -21,53 +21,90 @@
-function compileFPC(info, env) {
-    var compiler = new Compile(info, env);
-    compiler.supportsOptOutput = false;
+class PascalDemangler {
```
Oh fantastic! This is the first actual proper modern Javascript! Thank you!

mattgodbolt commented:
The demangling's fine where it is if you're happy; but I'd not be against a separate pascal-support library (or something).

mattgodbolt commented:
Should we hang on to merge until you've looked at the CPU issue, or are you happy as is? :)

RabsRincon commented:
> I think I should probably seperate the demangling class now into a different file.. any suggestions where i should put it?

The lib folder is probably the best choice

partouf commented:
Its definitely in some kind of infinite loop, I tried it twice, so better not merge yet.

But gotta get some sleep for the regular day job in a few hours :$

Sent from my iPhone

> On 28 Nov 2017, at 03:26, Matt Godbolt <notifications@github.com> wrote:
> 
> Should we hang on to merge until you've looked at the CPU issue, or are you happy as is? :)
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

mattgodbolt commented:
Haha thanks @partouf :) Rest up and we'll look at this tomorrow. Good night!

partouf commented:
This should be it.

RabsRincon commented:
So I'm seeing a bit of `==` instad of `===` which we'll eventually get some warnigns form the IDEs, but some of them do seem to be on pourpose, so I think we should leave them like this and finally merge.

It has been a lot of work, so we thank you for your efforts to bring Pascal to CE :)
Please do not hesitate in coming back, we'll love to hear back from you with more improvements if you want.

Note that in case you ever need to talk with us, check https://github.com/mattgodbolt/compiler-explorer#contact-us to know where

mattgodbolt commented:
ok @partouf - the beta site is up with the merged changes: https://godbolt.org/g/T8hta9

However it doesn't seem to be working, giving errors like:

```
Free Pascal Compiler version 3.0.2 [2017/02/05] for x86_64
Copyright (c) 1993-2016 by Florian Klaempfl and others
Note: Switching assembler to default source writing assembler
Target OS: Linux for x86-64
Compiling /tmp/compiler-explorer-compiler1171029-12-m82jz0.4iwb/prog.dpr
Compiling <source>
Assembling output
Assembling prog
Linking /tmp/compiler-explorer-compiler1171029-12-m82jz0.4iwb/prog
prog.dpr(1,53) Error: Error while linking
prog.dpr(1,53) Fatal: There were 1 errors compiling module, stopping
Fatal: Compilation aborted
Error: /opt/compiler-explorer/fpc-3.0.2.x86_64-linux/bin/ppcx64 returned an error exitcode
/usr/bin/ld: cannot find link.res: No such file or directory
Compiler exited with result code 1
```

Any ideas? @RabsRincon did you notice this at any point?

mattgodbolt commented:
Due to dreadful isolation on my part I've had to push this to the live site: sorry :( Please let me know if I can fix something quickly on my end!

partouf commented:
That's so weird, I'm guessing this is because of the -Xi parameter that was supposed to make fpc make use of an internal linker, so it wouldn't create this .sh file anymore to invoke /usr/bin/ld.
I'm guessing it created the file link.res as well "somewhere", which I didn't remove during testing on my system.
Can you see if it's solved by https://github.com/mattgodbolt/compiler-explorer/pull/662

partouf commented:
no link.res to be found in my local environment, but it still works...

mattgodbolt commented:
I've pushed to beta: https://pascal.godbolt.org/beta/ and as far as I can tell it is still not working...

I'm running the 1709 travis build which includes the merge:

```
ubuntu@ip-172-30-0-97 ~/compiler-explorer-image> ce --env beta status
Running in beta
Address          Instance Id          State      Type         ELB         Service     Version       
54.235.36.183    i-071380dce0ff1aee8  running    c5.large     healthy     running     1709 (master) 
```

https://travis-ci.org/mattgodbolt/compiler-explorer/builds/309090577
